File: 1140-stone-game-ii/1140-stone-game-ii.java
class Solution {
    int[][] dp;

    public int stoneGameII(int[] piles) {
        var n = piles.length;
        dp = new int[n][n];

        // calculate suffix sum (from right)
        for(var i=n-2; i>=0; i--) 
            piles[i] += piles[i+1];

        return score(piles, 1, 0);
    }

    private int score(int[] suffixSum, int m, int idx){
        if(idx + 2*m >= suffixSum.length) return suffixSum[idx];    // last player takes it all
        if(dp[m][idx] != 0) return dp[m][idx];

        var maxScore = 0;
        for(var i=1; i<= 2*m; i++){
            var take = suffixSum[idx] - suffixSum[idx+i];
            var leftover = suffixSum[idx+i] - score(suffixSum, Math.max(m, i), idx+i);
            maxScore = Math.max(maxScore, take + leftover);
        }

        dp[m][idx] = maxScore;
        return maxScore;
    }
}

File: 1140-stone-game-ii/1140-stone-game-ii.cs
public class Solution {
    int[] suffixSum;
    int[,] dp;

    public int StoneGameII(int[] piles) {
        var n = piles.Length;
        dp = new int[n,n];

        // calculate prefix sum (from right)
        suffixSum = new int[n+1];
        suffixSum[n] = 0;
        for(var i=n-1; i>=0; i--) suffixSum[i] = piles[i] + suffixSum[i+1];

        return Score(piles, 1, 0);
    }

    private int Score(int[] piles, int m, int idx){
        if(idx == piles.Length) return 0;
        if(idx == piles.Length - 1) return piles[idx];
        if(dp[m,idx] != 0) return dp[m,idx];

        var maxScore = 0;
        for(var i=1; i<= 2*m && idx+i<=piles.Length; i++){
            var take = suffixSum[idx] - suffixSum[idx+i];
            var leftover = suffixSum[idx+i] - Score(piles, Math.Max(m, i), idx+i);
            maxScore = Math.Max(maxScore, take + leftover);
        }

        dp[m,idx] = maxScore;
        return maxScore;
    }
}